build/lib/src/asset/reader.dart

```diff
@@ -36,11 +37,19 @@ abstract class AssetReader {
 
   /// Returns a [Digest] representing a hash of the contents of [id].
   ///
+  /// The digests should include the asset ID as well as the content of the
+  /// file, as some build systems may rely on the digests for two files being
+  /// different, even if their content is the same.
+  ///
   /// This should be treated as a transparent [Digest] and the implementation
   /// may differ based on the current build system being used.
   Future<Digest> digest(AssetId id) async {
-    var bytes = await readAsBytes(id);
-    return md5.convert(bytes);
+    var digestSink = AccumulatorSink<Digest>();
+    var bytesSink = md5.startChunkedConversion(digestSink);
+    bytesSink.add(await readAsBytes(id));
+    bytesSink.add([id.hashCode]);
```

Comment:
`ByteConversionSink` does take a `List<int>`, but afaik each of those ints must fit into a byte and we don't have a guarantee that `id.hashCode` does.

We could convert it, I think it would look like:

```dart
Uint8List intBytes(int number) {
  var result = new Uint8List(8);
  for (int i = 0; i < 8; i++) {
    result[7 - i] = number & 0xFF;
    number = number >> 8;
  }
  return result;
}
```

Alternatively do `bytesSink.add(utf8.encode('$id'))` or similar.